File: typescript/packages/mechanisms/tvm/src/exact/codec.ts
import { beginCell, loadMessageRelaxed } from "@ton/core";
import {
  ERR_EXACT_TVM_INVALID_SETTLEMENT_BOC,
  ERR_EXACT_TVM_INVALID_W5_ACTIONS,
  ERR_EXACT_TVM_INVALID_W5_MESSAGE,
  INTERNAL_SIGNED_OP,
  SEND_MODE_IGNORE_ERRORS,
  SEND_MODE_PAY_FEES_SEPARATELY,
} from "../constants";
import type { ParsedTvmSettlement } from "../types";
import { decodeBase64Boc, normalizeAddress } from "../codecs/common";
import { parseJettonTransfer } from "../codecs/jetton";
import { parseOutList } from "../codecs/w5";

export function parseExactTvmPayload(settlementBoc: string): ParsedTvmSettlement {
  let root;
  let message;
  try {
    root = decodeBase64Boc(settlementBoc);
    message = loadMessageRelaxed(root.beginParse());
  } catch (error) {
    throw Object.assign(new Error(ERR_EXACT_TVM_INVALID_SETTLEMENT_BOC), { cause: error });
  }

  if (message.info.type !== "internal") {
    throw new Error(ERR_EXACT_TVM_INVALID_SETTLEMENT_BOC);
  }

  const payer = normalizeAddress(message.info.dest);
  const body = message.body;
  const bodySlice = body.beginParse();

  if (bodySlice.remainingBits < 32) {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_MESSAGE);
  }
  const opcode = bodySlice.loadUint(32);
  if (opcode !== INTERNAL_SIGNED_OP) {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_MESSAGE);
  }

  const walletId = bodySlice.loadUint(32);
  const validUntil = bodySlice.loadUint(32);
  const seqno = bodySlice.loadUint(32);

  const hasActions = bodySlice.loadBit();
  const actionsCell = hasActions ? bodySlice.loadRef() : null;
  const actions = actionsCell ? parseOutList(actionsCell) : [];
  const hasExtraActions = bodySlice.loadBit();
  if (hasExtraActions) {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_ACTIONS);
  }

  if (actions.length !== 1 || actions[0].type !== "sendMsg") {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_ACTIONS);
  }

  const action = actions[0];
  if (
    action.outMsg.info.type !== "internal" ||
    !action.outMsg.info.dest ||
    action.outMsg.info.bounce !== true
  ) {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_ACTIONS);
  }

  const allowedModes = new Set([
    SEND_MODE_PAY_FEES_SEPARATELY,
    SEND_MODE_PAY_FEES_SEPARATELY + SEND_MODE_IGNORE_ERRORS,
  ]);
  if (!allowedModes.has(action.mode)) {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_ACTIONS);
  }

  const transfer = parseJettonTransfer(
    normalizeAddress(action.outMsg.info.dest),
    action.outMsg.body,
  );
  transfer.attachedTonAmount = action.outMsg.info.value.coins;

  const signature = bodySlice.loadBuffer(64);
  if (bodySlice.remainingBits || bodySlice.remainingRefs) {
    throw new Error(ERR_EXACT_TVM_INVALID_W5_MESSAGE);
  }

  const signedSliceBuilder = beginCell()
    .storeUint(opcode, 32)
    .storeUint(walletId, 32)
    .storeUint(validUntil, 32)
    .storeUint(seqno, 32);
  if (actionsCell) {
    signedSliceBuilder.storeBit(true).storeRef(actionsCell);
  } else {
    signedSliceBuilder.storeBit(false);
  }
  signedSliceBuilder.storeBit(false);

  return {
    payer,
    walletId,
    validUntil,
    seqno,
    settlementHash: root.hash().toString("hex"),
    body,
    signedSliceHash: signedSliceBuilder.endCell().hash(),
    signature,
    stateInit: message.init ?? null,
    transfer,
  };
}

File: python/x402/mechanisms/tvm/exact/codec.py
"""Exact TVM settlement payload codec."""

from __future__ import annotations

import base64

from ..codecs.common import normalize_address
from ..codecs.jetton import parse_jetton_transfer
from ..codecs.w5 import parse_out_list
from ..constants import (
    ERR_EXACT_TVM_INVALID_SETTLEMENT_BOC,
    ERR_EXACT_TVM_INVALID_W5_ACTIONS,
    ERR_EXACT_TVM_INVALID_W5_MESSAGE,
    SEND_MODE_IGNORE_ERRORS,
    SEND_MODE_PAY_FEES_SEPARATELY,
    W5_INTERNAL_SIGNED_OPCODE,
)
from ..types import ParsedTvmSettlement

try:
    from pytoniq_core import Cell
    from pytoniq_core.tlb.transaction import MessageAny
except ImportError as e:
    raise ImportError(
        "TVM mechanism requires pytoniq packages. Install with: pip install x402[tvm]"
    ) from e


def parse_exact_tvm_payload(settlement_boc: str) -> ParsedTvmSettlement:
    """Parse an exact TVM settlement payload into structured fields."""
    try:
        root = Cell.one_from_boc(base64.b64decode(settlement_boc))
        message = MessageAny.deserialize(root.begin_parse())
    except Exception as exc:
        raise ValueError(ERR_EXACT_TVM_INVALID_SETTLEMENT_BOC) from exc

    if not message.is_internal or message.info.dest is None:
        raise ValueError(ERR_EXACT_TVM_INVALID_SETTLEMENT_BOC)

    payer = normalize_address(message.info.dest)
    state_init = message.init
    body = message.body

    body_slice = body.begin_parse()
    opcode = body_slice.load_uint(32)
    if opcode != W5_INTERNAL_SIGNED_OPCODE:
        raise ValueError(ERR_EXACT_TVM_INVALID_W5_MESSAGE)

    wallet_id = body_slice.load_uint(32)
    valid_until = body_slice.load_uint(32)
    seqno = body_slice.load_uint(32)

    has_actions = body_slice.load_bit()
    actions = parse_out_list(body_slice.load_ref()) if has_actions else []
    has_extra_actions = body_slice.load_bit()
    if has_extra_actions:
        raise ValueError(ERR_EXACT_TVM_INVALID_W5_ACTIONS)

    if len(actions) != 1 or actions[0].type_ != "action_send_msg":
        raise ValueError(ERR_EXACT_TVM_INVALID_W5_ACTIONS)

    action = actions[0]
    if (
        not action.out_msg.is_internal
        or action.out_msg.info.dest is None
        or not action.out_msg.info.bounce
    ):
        raise ValueError(ERR_EXACT_TVM_INVALID_W5_ACTIONS)

    allowed_send_modes = {
        SEND_MODE_PAY_FEES_SEPARATELY,
        SEND_MODE_PAY_FEES_SEPARATELY + SEND_MODE_IGNORE_ERRORS,
    }
    if action.mode not in allowed_send_modes:
        raise ValueError(ERR_EXACT_TVM_INVALID_W5_ACTIONS)

    transfer = parse_jetton_transfer(
        jetton_wallet=normalize_address(action.out_msg.info.dest),
        body=action.out_msg.body,
    )
    transfer.attached_ton_amount = action.out_msg.info.value_coins

    signature = body_slice.load_bytes(64)
    if body_slice.remaining_bits or body_slice.remaining_refs:
        raise ValueError(ERR_EXACT_TVM_INVALID_W5_MESSAGE)

    signed_slice = body.begin_parse().copy()
    signed_slice.bits = signed_slice.bits[:-512]
    signed_slice_hash = signed_slice.to_cell().hash

    return ParsedTvmSettlement(
        payer=payer,
        wallet_id=wallet_id,
        valid_until=valid_until,
        seqno=seqno,
        settlement_hash=root.hash.hex(),
        body=body,
        signed_slice_hash=signed_slice_hash,
        signature=signature,
        state_init=state_init,
        transfer=transfer,
    )
